---
title: "haplotype_reconstruction_qtl-nf: Quality Control Report"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    code_folding: hide
date: "`r format(Sys.time(), '%m-%d-%Y')`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, dev = "png", fig.width = 8, fig.height = 6)
library(purrr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(plotly)
library(knitr)
library(qtl2)
library(broman)
library(qtlcharts)
summaryPlotTheme <- theme_bw(base_size = 14) + 
  theme(panel.grid = element_blank(),
        axis.title = element_text(colour = "black"))
args <- commandArgs(trailingOnly = TRUE)
options(scipen = 9999999)

```

## Load Data

```{r load data}

# QC output 1
load("QC_1.RData")
# load("/projects/compsci/vmp/USERS/widmas/haplotype_reconstruction_qtl-nf/projects/DO_ESC/geno_probs/QC_1.RData")

# cross object
load("preQC_cross.RData")
# load("/projects/compsci/vmp/USERS/widmas/haplotype_reconstruction_qtl-nf/projects/DO_ESC/geno_probs/preQC_cross.RData")

# x chromosome intensities
xint_file <- "chrXint.csv"
xint <- qtl2::fread_csv_numer(xint_file, transpose = TRUE)

# for testing
# xint_file <- "/projects/compsci/vmp/USERS/widmas/haplotype_reconstruction_qtl-nf/projects/do_oocyte/qtl2genos/chrXint.csv"
# xint <- qtl2::read_csv_numer(xint_file, transpose = TRUE)

# y chromosome intensities
yint_file <- "chrYint.csv"
yint <- qtl2::fread_csv_numer(yint_file, transpose=TRUE)

#for testing
# yint_file <- "/projects/compsci/vmp/USERS/widmas/haplotype_reconstruction_qtl-nf/projects/do_oocyte/qtl2genos/chrYint.csv"
# yint <- qtl2::read_csv_numer(yint_file, transpose = TRUE)

# crossover counts
load("nxos.RData")
# load("/projects/compsci/vmp/USERS/widmas/haplotype_reconstruction_qtl-nf/projects/DO_ESC/geno_probs/nxos.RData")
```

## Project Metrics

There were `r qtl2::n_ind(cross)` samples in the original experiment, comprising `r nrow(cross$covar[which(cross$covar$sex == "F"),])` females and `r nrow(cross$covar[which(cross$covar$sex == "M"),])` males. There were a total of `r sum(qtl2::n_mar(cross))` informative markers.

## Marker Metrics

### Missing Genotypes

```{r distribution of missing marker genotypes}
hist(percent_missing_marker, breaks = 100, 
     main = "Distribution of the proportion of samples missing each marker", 
     xlab = "Percent of samples missing genotype",
     ylab = "Number of genotypes")

remaining_markers_10pct <- length(which(percent_missing_marker > 10))/sum(n_mar(cross))
remaining_markers_5pct <- length(which(percent_missing_marker > 5))/sum(n_mar(cross))
```

Of all markers, `r paste0(round(remaining_markers_10pct*100),"%")` of markers were missing in 10% of samples and `r paste0(round(remaining_markers_5pct*100),"%")` in 5% of samples. We removed markers according to the latter threshold.

```{r showing marker depletion}
# cross with bad markers removed
cat("Original cross: \n")
qtl2::n_mar(cross)
cat("\nMarkers cleaned: \n")
qtl2::n_mar(working_cross)
```

## Sample Metrics

### Missing genotypes

```{r missing genotypes}
# label values with mouse ids
percent_missing_ind_cross <- qtl2::n_missing(working_cross, "ind", "prop")*100
labels <- paste0(names(percent_missing_ind_cross), " (", round(percent_missing_ind_cross), "%)")

# count number of samples at different missingness thresholds
missing_thresh <- unlist(lapply(c(1,5,10,20), function(x) length(percent_missing_ind_cross[percent_missing_ind_cross > x])))

# plot
qtlcharts::iplot(x = seq_along(percent_missing_ind_cross),
                 y = percent_missing_ind_cross, 
                 indID=labels,
                 chartOpts=list(xlab="Mouse", 
                                ylab="Percent missing genotype data",
                                ylim=c(0, 100))) %>% suppressMessages()

n <- names(sort(percent_missing_ind_cross, decreasing=TRUE))
iboxplot(log10(t(int[,n])+1), orderByMedian=FALSE, chartOpts=list(ylab="log10(SNP intensity + 1)"))
```

There were:

- `r missing_thresh[1]` samples missing genotypes for more than 1% of markers,
- `r missing_thresh[2]` samples missing more than 5%, 
- `r missing_thresh[3]` samples missing more than 10%, and
- `r missing_thresh[4]` samples missing more than 20%.

### Sample duplicates

Below is a list of sample pairs flagged as sample duplicates (if there were any). A histogram of the proportion of matching genotypes between all sample pairs is also provided.

```{r sample duplicates}
# default qtl2 report for sample duplicates
summary(cg)

# histogram
hist(cg[upper.tri(cg)], breaks=seq(0, 1, length=201),
     main="", yaxt="n", ylab="", xlab="Proportion matching genotypes")
rug(cg[upper.tri(cg)])
```

### Sex Checks

```{r sex checks}

# take average X chromosome probe intensity for non-null markers
# are any or all of the Y chromosome markers bad?
if(all(colnames(xint) %in% names(which(percent_missing_marker > 5)))){
  # just use what we have
  xint[is.na(xint)] <- mean(xint[!is.na(xint)])
  xint_ave <- rowMeans(xint[complete.cases(xint),])
} else if(any(colnames(xint) %in% names(which(percent_missing_marker > 5)))){
  # remove the bad markers
  xint <- xint[,-which(colnames(xint) %in% names(which(percent_missing_marker > 5)))]
  xint[is.na(xint)] <- mean(xint[!is.na(xint)])
  xint_ave <- rowMeans(xint[complete.cases(xint),])
} else {
  xint[is.na(xint)] <- mean(xint[!is.na(xint)])
  xint_ave <- rowMeans(xint[complete.cases(xint),])
}


# take average Y chromosome probe intensity for non-null markers
# are any or all of the Y chromosome markers bad?
if(all(colnames(yint) %in% names(which(percent_missing_marker > 5)))){
  # just use what we have
  yint[is.na(yint)] <- mean(yint[!is.na(yint)])
  yint_ave <- rowMeans(yint[complete.cases(yint),])
} else if(any(colnames(yint) %in% names(which(percent_missing_marker > 5)))){
  # remove the bad markers
  yint <- yint[,-which(colnames(yint) %in% names(which(percent_missing_marker > 5)))]
  yint[is.na(yint)] <- mean(yint[!is.na(yint)])
  yint_ave <- rowMeans(yint[complete.cases(yint),])
} else {
  yint[is.na(yint)] <- mean(yint[!is.na(yint)])
  yint_ave <- rowMeans(yint[complete.cases(yint),])
}

# put the Chr X and Y intensities together
ave_int <- data.frame(cbind(xint_ave, yint_ave)) %>%
  dplyr::mutate(sample = rownames(.))
ave_int <- cbind(ave_int,cross$covar[which(rownames(cross$covar) %in% rownames(ave_int)),])

# make cutoffs df
categories_df <- data.frame(c(0.25,0.5,0.25,0.5),
                            c(0.5,0.5,0.1,0.1),
                            c("M","XXY","XO","F"))
colnames(categories_df) <- c("X","Y","sex")

# sex palette
pal <- c("green","brown","pink","purple")
names(pal) <- categories_df$sex

# categorize samples
ave_int <- ave_int %>%
  dplyr::mutate(predicted_sex = dplyr::case_when(xint_ave > 0.45 & yint_ave < 0.15 ~ "F",
                                             xint_ave < 0.45 & yint_ave < 0.15 ~ "XO",
                                             xint_ave > 0.45 & yint_ave > 0.15 ~ "XXY",
                                             xint_ave < 0.45 & yint_ave > 0.15 ~ "M"),
                sample = rownames(.))

ave_int <- cbind(ave_int, percent_missing_ind_cross) %>%
  dplyr::mutate(percent_missing = paste0(round(percent_missing_ind_cross,2),"%"))

sex_check_plot <- ggplot() +
  geom_jitter(ave_int, mapping = aes(x = xint_ave, y = yint_ave, 
                                     fill = predicted_sex, 
                                     label = sample,
                                     label2 = percent_missing), 
              size = 3, shape = 21) +
  scale_fill_manual(values = pal) + 
  geom_hline(yintercept = 0.15) + 
  geom_vline(xintercept = 0.45) +
  labs(x = "Mean Chr X Probe Intensity",
       y = "Mean Chr Y Probe Intensity")%>%
  suppressWarnings()

# interactive sex check plot
plotly::ggplotly(sex_check_plot, tooltip = c("sample","percent_missing"))


pHet <- rowSums(working_cross$geno$X == 2)/rowSums(working_cross$geno$X != 0)
pHet <- cbind(ave_int, pHet)
pHet_plot <- ggplot() +
  geom_jitter(pHet, mapping = aes(x = xint_ave, y = pHet, 
                                  fill = predicted_sex, 
                                  label = sample,
                                  label2 = percent_missing), 
              size = 3, shape = 21) +
  scale_fill_manual(values = pal) +
  # geom_hline(yintercept = 0.15) + 
  # geom_vline(xintercept = 0.45) +
  labs(x = "Mean Chr X Probe Intensity",
       y = "Proportion Heterozygous Chr X Genotypes") %>%
  suppressWarnings()

# interactive pHet plot
plotly::ggplotly(pHet_plot, tooltip = c("sample","percent_missing"))
```

### Crossover Counts

```{r analyzing crossovers}
xo_counts <- data.frame(rowSums(nxo)) %>%
  dplyr::rename(nxos = `rowSums.nxo.`) %>%
  dplyr::mutate(sample = rownames(.)) %>%
  dplyr::left_join(., ave_int %>%
                     dplyr::select(sample, gen, predicted_sex, percent_missing_ind_cross)) %>%
  dplyr::mutate(percent_missing = paste0(round(percent_missing_ind_cross,2),"%"))

normal_count_plot <- ggplot(xo_counts) + 
  theme_bw() + 
  geom_jitter(mapping = aes(x = gen,
                            y = nxos,
                            label = sample,
                            label2 = percent_missing,
                            fill = predicted_sex), width = 0.1, height = 0, shape = 21, size = 2) +
  scale_fill_manual(values = pal) +
  theme(panel.grid.minor = element_blank()) + 
  labs(x = "Generation",
       y = "Number of crossovers")
  

if(working_cross$crosstype == "do"){
  xo_counts$expected_xos <- unlist(lapply(as.numeric(xo_counts$gen), function(x){
    expected = 300 + 24 * (x - 7)
    return(expected)
  }))
  
  oe_xo_count_plot <- ggplot(xo_counts) + 
    theme_bw() + 
    geom_jitter(mapping = aes(x = gen,
                            y = nxos-expected_xos,
                            label = sample,
                            label2 = percent_missing,
                            fill = predicted_sex), width = 0.2, shape = 21, size = 2) + 
    scale_fill_manual(values = pal) +
    theme(panel.grid.minor = element_blank()) + 
    labs(x = "Generation",
         y = "Observed crossovers - Expected crossovers")
}

ggplot(xo_counts) +
  theme_bw() +
  geom_histogram(mapping = aes(x = nxos-expected_xos))

# crossover count plot
plotly::ggplotly(normal_count_plot, tooltip = c("sample","percent_missing"))

# if possible plot the observed - expected plot
if(exists(x = "oe_xo_count_plot")){
  plotly::ggplotly(oe_xo_count_plot, tooltip = c("sample","percent_missing"))  
}

```

### Sample Status Update

Parsing the proportion of missing genotypes, predicted and recorded sexes, coat colors, and sample duplication, the following samples have been flagged for follow-up:

```{r sample status update}
# pull samples with duplicates
flagged_pairs <- apply(cg, 2, function(x){
  y <- x[which(x < 1)]
  z <- y[which(y > 0.9)]
  return(z)
})
sample_dups <- lapply(names(unlist(flagged_pairs)), function(x){
  strsplit(x, "[.]")[[1]]
  }) %>% unlist()
if(exists("FLAG_dup_df")){
  rm(FLAG_dup_df)
}
if(!is.null(sample_dups)){
 FLAG_dup_df <- data.frame(sample_dups, "sample duplicate") %>%
  `colnames<-`(c("sample","flag_reason"))
}
FLAG_dup_df <- FLAG_dup_df[!duplicated(FLAG_dup_df),]


# pull samples with many missing genotypes
FLAG_pct_miss <- data.frame(percent_missing_ind_cross[percent_missing_ind_cross > 5]) %>%
  dplyr::mutate(sample = rownames(.)) %>%
  `colnames<-`(c("notes","sample")) %>%
  dplyr::mutate(notes = round(notes, 2),
                notes = paste0(notes,"% of markers missing genotypes"))
rownames(FLAG_pct_miss) <- NULL
FLAG_pct_miss$flag_reason <- "missing genotypes"

# pull mismatched assigned and predicted sex
FLAG_sex_mismatch_df <- ave_int %>%
  dplyr::filter(sex != predicted_sex) %>%
  dplyr::mutate(notes = paste0("predicted sex: ",predicted_sex,"; recorded sex: ", sex),
                flag_reason = "sex mismatch") %>%
  dplyr::select(sample, notes, flag_reason)
rownames(FLAG_sex_mismatch_df) <- NULL

# pull samples with wild crossover counts
FLAG_xo <- xo_counts %>%
  dplyr::select(sample, nxos) %>%
  dplyr::filter(nxos > 2000 | nxos < 500) %>%
  dplyr::mutate(notes = paste0(nxos, " crossovers; flagged as too low or high"),
                flag_reason = "crossover counts") %>%
  dplyr::select(-nxos)

# combine all the reports
flag_report <- lapply(ls(pattern = "FLAG"), function(x) get(x)) %>%
  Reduce(dplyr::bind_rows,.) %>%
  dplyr::select(sample, flag_reason, notes) %>%
  dplyr::arrange(sample) %>%
  dplyr::mutate(notes = dplyr::if_else(flag_reason == "sample duplicate",
                                       true = "T", false = notes))
rownames(flag_report) <- NULL

# tidy the sample flag report?
wide_flag_report <- flag_report %>%
  tidyr::pivot_wider(names_from = flag_reason,
                     values_from = notes)

# display flagged samples
DT::datatable(data = wide_flag_report, 
              class = 'cell-border stripe', 
              options = list(pageLength = 20), 
              rownames = FALSE)

# write flagged samples
write.csv(flag_report, quote = F, row.names = F, file = "sample_QC_result.csv")
```
